experimental/iterators/include/operators/reduce_by_key.h

```diff
@@ -0,0 +1,113 @@
+//===-- operators/reduce_by_key.h - ReduceByKeyOperator ---------*- C++ -*-===//
+///
+/// \file
+/// This file contains the definition of the `ReduceByKeyOperator class` as
+/// well as related helpers.
+///
+//===----------------------------------------------------------------------===//
+#ifndef OPERATORS_REDUCE_BY_KEY_H
+#define OPERATORS_REDUCE_BY_KEY_H
+
+#include <optional>
+#include <tuple>
+#include <unordered_map>
+
+#include "utils/tuple.h"
+
+/// Groups the input tuples by key and reduces the tuples within each group.
+///
+/// This operator consumes the tuples produced by its upstream operator, groups
+/// them based on their key attributes (a given number of attributes starting
+/// at the beginning of each tuple), and combines (aka reduces or folds) the
+/// tuples of each group into a single tuple using the given reduce function.
+/// If the upstream operator does not produce any tuple, neither does this
+/// operator. The reduce function must accept two arguments that correspond
+/// both to the tail of the tuple produced by the upstream operator (i.e., the
+/// remainder after the key attributes) and return a tuple of the same type.
+template <typename UpstreamType, typename ReduceFunctionType,
+          std::size_t kNumKeyAttributes>
+class ReduceByKeyOperator {
+public:
+  using OutputTuple = typename UpstreamType::OutputTuple;
+  using ReturnType = std::optional<OutputTuple>;
+
+  /// Constructs a new `ReduceByKeyOperator` that holds a reference on its
+  /// upstream operator and a copy of the provided `reduceFunction`.
+  explicit ReduceByKeyOperator(UpstreamType *const upstream,
+                               ReduceFunctionType reduceFunction)
+      : upstream(upstream), reduceFunction(std::move(reduceFunction)),
+        currentResultIt(result.begin()), endResultIt(result.end()) {}
+
+  /// Opens the upstream operator.
+  void open() { upstream->open(); }
+
+  /// In the first call to this function, the tuples produced by the upstream
+  /// operator are consumed and each tuple is combined with an intermediate
+  /// internal aggregate of all previous tuples with the same key (i.e., in the
+  /// same group). This first and the remaining calls return one fully reduced
+  /// tuple per call. When no tuples remain in the result, "end-of-stream" is
+  /// returned.
+  ReturnType computeNext() {
+    // This condition is true either if (1) this is the first call to
+    // computeNext (cf how the two iterators are initialized in the
+    // constructor) or if (2) it is the last, i.e., after all results produced
+    // by this operator have already been returned.
+    // Note that there is only a single test in the common case.
+    if (currentResultIt == endResultIt) {
+      // (2) All tuples have been returned; signal "end-of-stream"
+      if (hasConsumedUpstream) {
```

Comment:
in LLVM/MLIr we omit trivial braces, fine to leave for now unless you expect some of this code to survive longer term.

Reply:
Fixed in files added in this PR. A next PR will fix it in the remaining files.

Reply:
sg!